src/main/java/org/opensearch/knn/index/vectorvalues/KNNVectorValuesFactory.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.vectorvalues;
+
+import org.apache.lucene.index.DocsWithFieldSet;
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.search.DocIdSetIterator;
+
+import java.util.Map;
+
+/**
+ * A factory class that provides various methods to create the {@link KNNVectorValues}.
+ */
+public final class KNNVectorValuesFactory {
```

Comment:
there is no use of Vector datatype parameter to be passed as the factory class is providing concrete functions. The problem with the suggested methods is introduction of `?` in the return type. Which make it very hard for caller to know what is the return type. Hence its not needed.

Reply:
If I have 10 child classes , By using the implemented way for factory pattern , I can have 10 methods which return instance of 10 child objects individually  and that's factory pattern?

We have something called generics in Java , that helps creating generic methods , here is how we can do that 

```
package org.opensearch.knn.index.vectorvalues;

public enum VectorType {
    FLOAT,
    BYTE,
    BINARY
}
```
package org.opensearch.knn.index.vectorvalues;

import org.apache.lucene.search.DocIdSetIterator;

import java.util.Map;

public class KNNVectorValuesFactory {

    public static <T> KNNVectorValues<T> getVectorValues(VectorType vectorType, DocIdSetIterator docIdSetIterator) {
        KNNVectorValuesIterator vectorValuesIterator = new KNNVectorValuesIterator.DocIdsIteratorValues(docIdSetIterator);

        switch (vectorType) {
            case FLOAT:
                return (KNNVectorValues<T>) new KNNFloatVectorValues(vectorValuesIterator);
            case BYTE:
                return (KNNVectorValues<T>) new KNNByteVectorValues(vectorValuesIterator);
            case BINARY:
                return (KNNVectorValues<T>) new KNNBinaryVectorValues(vectorValuesIterator);
            default:
                throw new IllegalArgumentException("Unknown vector type: " + vectorType);
        }
    }

    public static <T> KNNVectorValues<T> getVectorValues(VectorType vectorType, DocsWithFieldSet docIdWithFieldSet, Map<Integer, T> vectors) {
        KNNVectorValuesIterator vectorValuesIterator = new KNNVectorValuesIterator.FieldWriterIteratorValues<>(docIdWithFieldSet, vectors);

        switch (vectorType) {
            case FLOAT:
                return (KNNVectorValues<T>) new KNNFloatVectorValues(vectorValuesIterator);
            case BYTE:
                return (KNNVectorValues<T>) new KNNByteVectorValues(vectorValuesIterator);
            case BINARY:
                return (KNNVectorValues<T>) new KNNBinaryVectorValues(vectorValuesIterator);
            default:
                throw new IllegalArgumentException("Unknown vector type: " + vectorType);
        }
    }
}


Here is how Client can call, 

```
DocIdSetIterator iterator = ...; // Your iterator instance
KNNVectorValues<float[]> floatVectors = KNNVectorValuesFactory.getVectorValues(VectorType.FLOAT, iterator);
KNNVectorValues<byte[]> byteVectors = KNNVectorValuesFactory.getVectorValues(VectorType.BYTE, iterator);

// For the case with DocsWithFieldSet and a map of vectors
DocsWithFieldSet docsWithFieldSet = ...;
Map<Integer, float[]> floatVectorMap = ...;
KNNVectorValues<float[]> floatVectorsWithMap = KNNVectorValuesFactory.getVectorValues(VectorType.FLOAT, docsWithFieldSet, floatVectorMap);

```

Reply:
+1 with current code. 

When you call `KNNVectorValuesFactory.getVectorValues(VectorType.BINARY, iterator);` it is hard to know if its return type is `KNNVectorValues<float[]>` or `KNNVectorValues<byte[]>`. You need to go inside the method and check what is its return value. 
By having `public static KNNVectorValues<byte[]> getBinaryVectorValues(final DocIdSetIterator docIdSetIterator)`, it is more explicit and you just know what it will return without going inside the method and check the actual return value.

Reply:
Returnning `?` type is not ideal in static type system where you cannot catch the issue during compile time but only runtime.

Reply:
well when u call KNNVectorValuesFactory.getVectorValues(VectorType.BINARY, iterator) , it says that vectorr value is type Binary as you hve already speicified. while calling , I don't see any reason on confusion 

So here was origianl thing 

KNNVectorValues<float[]> x = KNNVectorValuesFactory.getFloatVectorValue()
vs 
KNNVectorValues<float[]> floatVectorsWithMap = KNNVectorValuesFactory.getVectorValues(VectorType.FLOAT, docsWithFieldSet, floatVectorMap);

where is the confusion?

Reply:
> Returnning `?` type is not ideal in static type system where you cannot catch the issue during compile time but only runtime.

I slightly disagree on this , this is from Java Creator Effective Java (3rd Edition)  Joshua Bloch

Item number :- 29 ,  Consider typesafe heterogeneous containers and Favor generic methods, 
Bloch advocates for using generic methods to increase flexibility and reusability. He explains how generic methods can help avoid casting and improve type safety.

I don't find any documents which says generics is bad  and we should not use , rather I find useful to use that when we have lots of similar code to make it consistent.

Reply:
You can compile following code as well and you will find out that it break somewhere during run time.
```
KNNVectorValues<float[]> floatVectorsWithMap = KNNVectorValuesFactory.getVectorValues(VectorType.BINARY, docsWithFieldSet, floatVectorMap);
```

I am not saying generic is bad. However, you need to use is properly for it to be good. Returning ? does not help to avoid type casting and improve type safety.

Reply:
> ```
> public static <T> KNNVectorValues<T> getVectorValues(VectorType vectorType, DocIdSetIterator docIdSetIterator) {
>     KNNVectorValuesIterator vectorValuesIterator = new KNNVectorValuesIterator.DocIdsIteratorValues(docIdSetIterator);
> 
>     switch (vectorType) {
>         case FLOAT:
>             return (KNNVectorValues<T>) new KNNFloatVectorValues(vectorValuesIterator);
>         case BYTE:
>             return (KNNVectorValues<T>) new KNNByteVectorValues(vectorValuesIterator);
>         case BINARY:
>             return (KNNVectorValues<T>) new KNNBinaryVectorValues(vectorValuesIterator);
>         default:
>             throw new IllegalArgumentException("Unknown vector type: " + vectorType);
>     }
> }
> 
> public static <T> KNNVectorValues<T> getVectorValues(VectorType vectorType, DocsWithFieldSet docIdWithFieldSet, Map<Integer, T> vectors) {
>     KNNVectorValuesIterator vectorValuesIterator = new KNNVectorValuesIterator.FieldWriterIteratorValues<>(docIdWithFieldSet, vectors);
> 
>     switch (vectorType) {
>         case FLOAT:
>             return (KNNVectorValues<T>) new KNNFloatVectorValues(vectorValuesIterator);
>         case BYTE:
>             return (KNNVectorValues<T>) new KNNByteVectorValues(vectorValuesIterator);
>         case BINARY:
>             return (KNNVectorValues<T>) new KNNBinaryVectorValues(vectorValuesIterator);
>         default:
>             throw new IllegalArgumentException("Unknown vector type: " + vectorType);
>     }
> }
> ```

@Vikasht34 in your implementation why are we repeating the code mentioned below? A simple method getVectorType which has method definition mentioned below will solve the purpose.
> ```
>         case FLOAT:
>             return (KNNVectorValues<T>) new KNNFloatVectorValues(vectorValuesIterator);
>         case BYTE:
>             return (KNNVectorValues<T>) new KNNByteVectorValues(vectorValuesIterator);
>         case BINARY:
>             return (KNNVectorValues<T>) new KNNBinaryVectorValues(vectorValuesIterator);
>         default:
>             throw new IllegalArgumentException("Unknown vector type: " + vectorType);
> ```

Reply:
by seeing all the comments I will go ahead and first try this:

```
KNNVectorValues<float[]> x = KNNVectorValuesFactory.getFloatVectorValue(docsWithFieldSet, floatVectorMap)
vs
KNNVectorValues<float[]> floatVectorsWithMap = KNNVectorValuesFactory.getVectorValues(VectorType.FLOAT, docsWithFieldSet, floatVectorMap);
```
I am not particularly happy with sending VectorDataType in params, but since its not a one way door, and very much internal to k-NN hence I will change it to use switch, only if we can avoid `?` in the method signature.

Reply:
I am not happy with this change becase there is no difference between `?` and `T` but won't push further if it is already decided.